### PR TITLE
Delete IAM resources if associated instance is not running

### PR DESCRIPTION
**Pull request by sallyom**
@dgoodwin @joelddiaz I noticed if a cluster doesn't complete successfully, there is likely  IAM roles/policies/profiles that are not associated with a running instance (bc they are created in the install process b4 the instance).  The deprovisioner was dependent upon an instance running to delete the IAM resources.  I've added the function of deleting IAM resources for the specific profiles we know to be created with the installer. `${CLUSTER_NAME-master-profile` `${CLUSTER_NAME-worker-profile` and `${CLUSTER_NAME-bootstrap-profile`
have to now pass `--cluster-name mycluster` along with the tags
I've tested different scenerios:  failed launch/successful launch/single/multiple masters  this works well

### CODE REVIEW

File: contrib/pkg/aws_tag_deprovision/aws_tag_deprovision.go [hunk 1]
```diff
@@ -697,9 +682,43 @@ func deleteInstanceProfile(instanceProfileID *string, iamClient *iam.IAM, logger
...
+			err = deleteInstanceProfile(profile.InstanceProfile.InstanceProfileId, iamClient, logger)
+			if err != nil {
+				logger.Debugf("error deleting instance profile: %v", err)
+				return false, nil
+			}
+		}
+	}
+	return true, nil
+}
+
```
Comment (sallyom) [inline contrib/pkg/aws_tag_deprovision/aws_tag_deprovision.go]: there could be a way so we don't have to depend on the hardcoded installer_type above.. or a better name than `installer_type`..

File: contrib/pkg/aws_tag_deprovision/aws_tag_deprovision.go [hunk 2]
```diff
@@ -252,25 +258,6 @@ func tagsToMap(tags interface{}) (map[string]string, error) {
...
-		if *ip.Arn == *ipARN {
-			return ip.InstanceProfileId, nil
-		}
-	}
-
-	// if we didn't find it, that means it may already have been deleted
-	return nil, nil
-}
-
 // lbToAWSObjects will create awsObjectWithTags given a list of load balancers
```
Comment (sallyom) [inline contrib/pkg/aws_tag_deprovision/aws_tag_deprovision.go]: That function wasn't being used, so I've cleaned it up here..

File: .gitignore [hunk 3]
```diff
@@ -4,8 +4,6 @@
 *.dll
 *.so
 *.dylib
-<<<<<<< HEAD
-=======
```
Comment (sallyom) [inline .gitignore]: that isn't supposed to be there, right?

File: contrib/pkg/aws_tag_deprovision/aws_tag_deprovision.go [hunk 4]
```diff
@@ -697,9 +681,88 @@ func deleteInstanceProfile(instanceProfileID *string, iamClient *iam.IAM, logger
...
+	logger.Debugf("deleting instance profile: %v", profileName)
+	_, err = iamClient.DeleteInstanceProfile(&iam.DeleteInstanceProfileInput{
+		InstanceProfileName: &profileName,
+	})
+	if err != nil && err.(awserr.Error).Code() != iam.ErrCodeNoSuchEntityException {
+		logger.Debugf("error deleting instance profile %s: %v", profileName, err)
+		return err
+	}
+	if err == nil {
+		logger.Debugf("deleted profile %s", profileName)
```
Comment (joelddiaz) [inline contrib/pkg/aws_tag_deprovision/aws_tag_deprovision.go]: Infof() for non-error

File: contrib/pkg/aws_tag_deprovision/aws_tag_deprovision.go [hunk 5]
```diff
@@ -697,9 +681,88 @@ func deleteInstanceProfile(instanceProfileID *string, iamClient *iam.IAM, logger
...
+	// now we can delete role
+	logger.Debugf("deleted policies from role %s", roleName)
+	deleteRoleInput := iam.DeleteRoleInput{}
+	deleteRoleInput.RoleName = &roleName
+	if _, err := iamClient.DeleteRole(&deleteRoleInput); err != nil && err.(awserr.Error).Code() != iam.ErrCodeNoSuchEntityException {
+		logger.Debugf("error deleting role %s: %v", roleName, err)
+		return err
+	}
+	if err == nil {
+		logger.Debugf("deleted role %s", roleName)
```
Comment (joelddiaz) [inline contrib/pkg/aws_tag_deprovision/aws_tag_deprovision.go]: Infof() for non-error

File: contrib/pkg/aws_tag_deprovision/aws_tag_deprovision.go [hunk 6]
```diff
@@ -697,9 +681,88 @@ func deleteInstanceProfile(instanceProfileID *string, iamClient *iam.IAM, logger
...
+
+// deleteIAMresources will delete any IAM resources created by the installer that are not associated with a running instance
+// Currently openshift/installer creates 3 roles per cluster, 1 for master|worker|bootstrap and identified by the
+// cluster name used to install the cluster.
+func deleteIAMresources(session *session.Session, filter awsFilter, clusterName string, logger log.FieldLogger) (bool, error) {
+	logger.Debugf("Deleting IAM resources")
+	defer logger.Debugf("Exiting deleting IAM resources")
+	installer_type := []string{"master", "worker", "bootstrap"}
+	for _, t := range installer_type {
+		roleName := fmt.Sprintf("%s-%s-role", clusterName, t)
```
Comment (joelddiaz) [inline contrib/pkg/aws_tag_deprovision/aws_tag_deprovision.go]: can we put a comment here explaining the expected naming scheme for the various IAM objects?

File: contrib/pkg/aws_tag_deprovision/aws_tag_deprovision.go (hunk 4 above)
Comment (sallyom) [reply contrib/pkg/aws_tag_deprovision/aws_tag_deprovision.go]: fixed

File: contrib/pkg/aws_tag_deprovision/aws_tag_deprovision.go (hunk 5 above)
Comment (sallyom) [reply contrib/pkg/aws_tag_deprovision/aws_tag_deprovision.go]: fixed

**Pull request merged**